subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputs.java

```diff
@@ -84,11 +84,12 @@
      * <p>Registers an input property for this task. This value is persisted when the task executes, and is compared
      * against the property value for later invocations of the task, to determine if the task is up-to-date.</p>
      *
-     * <p>The given value for the property must be Serializable, so that it can be persisted. It should also provide a
-     * useful {@code equals()} method.</p>
+     * <p>The given value must be a simple value, like a String or Integer, or serializable. For complex values,
+     *  Gradle compares the serialized forms for detecting changes and the {@code equals()} method is ignored.
      *
-     * <p>You can specify a closure or {@code Callable} as the value of the property. In which case, the closure or
-     * {@code Callable} is executed to determine the actual property value.</p>
+     * <p>You can specify a {@link org.gradle.api.provider.Provider}, {@link java.util.concurrent.Callable}, Groovy closure,
+     * or Kotlin function as the value of the property. In which case, the function is executed to determine the actual
+     * property value.</p>
```

Comment:
Maybe we should split those things in two parts. How about
```suggestion
     * <p>If the value is not known when registering the input, you should register a {@link org.gradle.api.provider.Provider} which Gradle evaluates lazily.
     * You can specify a {@link java.util.concurrent.Callable}, a Groovy closure,
     * or a Kotlin function as the value of the property. In which case, the function is executed to determine the actual
     * property value. You should prefer using a {@link org.gradle.api.provider.Provider}.</p>
```

Reply:
Sounds good to me as well.

Reply:
How about:
```suggestion
     * <p>If the value is not known when registering the input, a {@link org.gradle.api.provider.Provider} can be
     * passed instead. Gradle will then resolve the provider at the latest possible time in order to determine the actual
     * property value.</p>
```